### ISSUE TITLE
Alter table options

### ISSUE THREAD

**Issue by killme2008** [GreptimeTeam/greptimedb#1084]
### What type of enhancement is this?

API improvement, Configuration

### What does the enhancement do?

For example, modify the table's `TTL` to 30 days:
```sql
ALTER TABLE monitor MODIFY TTL '30d';
```
Not all of the table options can be modified. The `regions` option can't be changed right now. Maybe we can manage a white list that contains alterable options.

### Implementation challenges

How to make the change takes effect dynamically? I am not sure. I think we can implement a observer pattern for it.

**Comment by killme2008** [comment]
After discussing and investigating, decide to use the following syntax:

```sql
alter table set options(ttl='30d', write_buffer_size='32mb');
```

The syntax for options is the same as that for creating tables. The other keywords in the `options` place can be used to set other table features.

**Comment by NiwakaDev** [comment]
I guess that we don't allow changing `storage` option.

**Comment by killme2008** [comment]
> I guess that we don't allow changing `storage` option.

Yes, of course.  I've done some preliminary work in #210 .

**Comment by chienguo** [comment]
@killme2008 Hi! why did you stop this work? I am also interested in this issue, and I also noticed the discussion under  https://github.com/GreptimeTeam/greptimedb/issues/4394.  Seems that alter table options and alter database options can be done parallelly. Can I have a try on this issue?

**Comment by evenyag** [comment]
Feel free to try it. Could you please share your design and steps for implementing this feature?

**Comment by chienguo** [comment]
here is my rough design:
- [ ] add `AlterKind::ChangeOptions`, record the option names that are about to change and their corresponding value
- [ ] for all the places that deal with AlterKind, add support for `ChangeOptions`, like `alter_expr_to_request`, `create_proto_alter_kind` to create region proto, `builder_with_alter_kind` to construct a `TableMetaBuilder` actually to set table options for the upper `AlterTableProcedure` to use
- [ ] for the table options white list, for now, I can only think of doing it in a hardcode way
- [ ] modify parser to support input like this `alter table set options(ttl='30d', write_buffer_size='32mb');`

And where the proto file to generate `greptime.v1.meta.rs` locates? I cannot find it

**Comment by evenyag** [comment]
> And where the proto file to generate `greptime.v1.meta.rs` locates? I cannot find it

Are you looking for this?
https://github.com/GreptimeTeam/greptime-proto/tree/main/proto/greptime/v1

What's more, we also need to support altering options in the region engine and update the option of the region.

**Comment by chienguo** [comment]
Since tables are split and stored in regions, should altering table options result in region options being updated cascadingly? And besides that, do we also need to support altering region options directly?

**Comment by zyy17** [comment]
> here is my rough design:
> 
> * [ ]  add `AlterKind::ChangeOptions`, record the option names that are about to change and their corresponding value
> * [ ]  for all the places that deal with AlterKind, add support for `ChangeOptions`, like `alter_expr_to_request`, `create_proto_alter_kind` to create region proto, `builder_with_alter_kind` to construct a `TableMetaBuilder` actually to set table options for the upper `AlterTableProcedure` to use
> * [ ]  for the table options white list, for now, I can only think of doing it in a hardcode way
> * [ ]  modify parser to support input like this `alter table set options(ttl='30d', write_buffer_size='32mb');`
> 
> And where the proto file to generate `greptime.v1.meta.rs` locates? I cannot find it

I'm also very interested in this issue. Are you considering implementing it? If so, I can assign it to you.

**Comment by chienguo** [comment]
@zyy17 Yes, indeed, could you please assign it to me?

**Comment by evenyag** [comment]
It's recommended to support altering region options first. We don't store the option in the manifest so we can update the [`version`](https://github.com/GreptimeTeam/greptimedb/blob/d628079f4c85c0d3e62991c9d2abb71070e51e31/src/mito2/src/region/version.rs#L218-L244) of the region here if the request is `ModifyOptions`.
https://github.com/GreptimeTeam/greptimedb/blob/d628079f4c85c0d3e62991c9d2abb71070e51e31/src/mito2/src/worker/handle_alter.rs#L115-L126

**Comment by WenyXu** [comment]
BTW, maybe we don't need the `options` keywords, the stmt without `options` is more concise.

```sql
ALTER TABLE table_name SET ( storage_parameter [= value] [, ... ] )
```
```sql
ALTER TABLE table_name UNSET ( storage_parameter [, ... ] )
```

**Comment by killme2008** [comment]
> BTW, maybe we don't need the `options` keywords, the stmt without `options` is more concise.
> 
> ```sql
> ALTER TABLE table_name SET ( storage_parameter [= value] [, ... ] )
> ```
> 
> ```sql
> ALTER TABLE table_name UNSET ( storage_parameter [, ... ] )
> ```

We may support altering other features in the future, so please keep the `options` keyword.

**Comment by WenyXu** [comment]
> We may support altering other features in the future, so please keep the `options` keyword.

The `ALTER TABLE table_name SET (ttl='30d', write_buffer_size='32mb')`, can be considered as starting with `LeftParen` (`(`), so I think we can distinguish between `SET (` and `SET STH` in the future.

**Comment by killme2008** [comment]
> > We may support altering other features in the future, so please keep the `options` keyword.
> 
> The `ALTER TABLE table_name SET (ttl='30d', write_buffer_size='32mb')`, can be considered as starting with `LeftParen` (`(`), so I think we can distinguish between `SET (` and `SET STH` in the future.

It's a bad practice to distinguish something with `(`.

**Comment by Arshil-Akkala** [comment]
Are these changes merged yet, if not.. is there an estimate? @killme2008

**Comment by killme2008** [comment]
@Arshil-Akkala #4848 has been merged and will be included in v0.10, releasing on September 15th. Before that, you can use the weekly built.
